Update Folder and Project modules' structure

AlexBMes opened the pull request:
Updated the `google_folder` and `google_project` modules for better management of multiple levels of folders and projects. And added a README for each of the modules. 
Haven't fully tested changes as org is required, but did some mock tests to check changes to locals configuration. Will use deployment tests to confirm no errors. 
Changes include:
* Changed structure of `folders/project_properties`local variables. Previously was:
  ```terraform
  {
    "1" = {
      display_name = "production"
      folder_iam = {
        ...
      }
    }
    "2" = {
      display_name = "test"
      folder_iam = {
        ...
      }
    }
  }
  ```
  but changed to use `display_name` (or `name` for projects`) as the key. So:
  ```terraform
  {
    "production" = {
      display_name = "production"
      folder_iam = {
        ...
      }
    }
    "test" = {
      display_name = "test"
      folder_iam = {
        ...
      }
    }
  }
  ```
* added `folders/project_iam` local variables, to be used in the IAM policy configuration. Previously if the `folders/project_iam` values were not configured in the yaml file, the IAM policy would be set but would be empty, potentially removing all IAM configuration for a project.  
`locals.folder_iam` will now have the structure `<folder display name> = <map of IAM policy>` only if the `folder_iam` attribute is present in the yaml file. Meaning terraform won't manage the policy unless it specified.
* added a `folder_names` output, which outputs a map of folder names created by the module, e.g.:
  ```terraform
  names = {
    "bash" = "folders/1234567"
    "java" = "folders/1234568"
    "python" = "folders/1234569"
    "react" = "folders/1234560"
  }
  ``` 
  So individual folder ID's can then be retrieved using `module.<module_name>.folder_names["<folder-name"]`
* added a `parent_folder` variable to both the project and folder modules, allowing sub folders and projects to created dynamically based on folders also created through terraform. This is needed because otherwise, setting in the yaml file would require a script to dynamically update the `parent_id`/`folder_id` to a folder just created. Example usage (included in documentation):
  ```terraform
  module core_folders {
    source      = "github.com/mesoform/Multi-Cloud-Platform-Foundations//Google/resource-manager/folder"
    projects_yml = "./core_folders.yaml"
  }
  module dev_folders {
    source      = "github.com/mesoform/Multi-Cloud-Platform-Foundations//Google/resource-manager/folder"
    projects_yml = "./dev_folders.yaml"
    parent_folder = module.core_folders.folder_names["development"]
  }
  module dev_projects {
    source      = "github.com/mesoform/Multi-Cloud-Platform-Foundations//Google/resource-manager/project"
    projects_yml = "./dev_projects.yaml"
    parent_folder = module.core_folders.folder_names["development"]
  }
  ```

gbmeuk commented on Google/resource-manager/folder/locals.tf:
```diff
@@ -3,8 +3,13 @@ locals {
-  folder_properties = {for key, value in lookup (local.folder_components, "folders", {}) : 
-    key => {for name, content in value : name => value["${name}"]}
+  parent_id = var.parent_folder == null ? lookup(local.folder_components, "parent_id", null) : var.parent_folder
+  folder_properties = {for key, value in lookup (local.folder_components, "folders", {}) :
```
I think this is still wrong because `folders` is a list of maps

gbmeuk commented on Google/resource-manager/folder/locals.tf:
```diff
@@ -3,8 +3,13 @@ locals {
   folders = try(yamldecode(local.folders_yml), {})
   folder_components       = try(lookup(local.folders, "components", {}), lookup(local.folders, "components"))
```
We don't actually have `specs` https://github.com/mesoform/Multi-Cloud-Platform#adapter-top-level-attributes

gbmeuk reviewed (changes requested):
I like the idea around `var.parent_id` but I think this module doesn't really meet our MCCF standard properly.

AlexBMes commented:
Updated the module to fit MCCF format. An example configuration would now be:
```yaml
components:
  common:
    org_id: 12345678901
    #folder_id: 12345678
    billing_account: 12345-12345-12345
    skip_delete: true
    labels: 
      key-1: value-1
    auto_create_network: false

  specs: 
    staging-sandbox:
      name: Staging Sandbox
      project_id: project-id-123456
      skip_delete: true
      labels:
        key-1: overwrite-value
        key-2: value-2
      auto_create_network: true
      project_iam:
        - role: "roles/viewer"
          members:
            - "user:info@kvit.pub"
        - role: "roles/viewer"
          members:
            - "user:user@kvit.pub"
          condition:
            title: "expires_after_2019_12_31"
            description: "Expiring at midnight of 2019-12-31"
            expression: "request.time < timestamp(\"2020-01-01T00:00:00Z\")"
    test-project:
      project_id: test-project-123

```

Also added new unit tests, and updated deployment tests

gbmeuk commented on Google/resource-manager/folder/README.md:
```diff
@@ -0,0 +1,74 @@
+          members:
+            - "user:info@example.com"
+    department-2:
+      display_name: "Department 2"
```
This isn't valid: https://cloud.google.com/resource-manager/reference/rest/v2/folders

AlexBMes replied:
The `display_name`? I don't see what isn't valid. Tested a deployment with that display name and it worked.
